AP_Scripting: add Lua CRSF Interface

qqqlab opened the pull request:
Add two functions to Lua scripting to receive and send telemetry data via a CRSF/ExpressLRS receiver, see included .md file for details.

Commit:
AP_Scripting: add Lua CRSF Interface

andyp1per commented on libraries/AP_RCProtocol/AP_RCProtocol_CRSF.cpp:
```diff
@@ -384,6 +384,7 @@ bool AP_RCProtocol_CRSF::decode_crsf_packet()
             process_link_stats_tx_frame((uint8_t*)&_frame.payload);
             break;
         default:
+            process_lua_pop();
```
You need to get a custom frame ID registered with TBS rather than just assuming everything is LUA. I can do this for you if you would like.
Needs to be surrounded with #if AP_SCRIPTING_ENABLED

qqqlab replied:
Hi Andy,

Thanks for your prompt feedback. I've been using Ardupilot since APM2.5 with great joy, and it a pleasure to finally be able to make a small contribution.

My thought behind using 'default' instead of a specific custom frame ID was to give Lua the ability handle anything that is not handled by AP. Similar to what is happening on the EdgeTX side.

But there are of course also arguments that speak for using specific ID, for example to prevent a potential flooding of Lua with unwanted messages.

Let's go with the specific ID, I've added a placeholder for now: CRSF_FRAMETYPE_AP_LUA = 0x81.

Please get a custom frame ID registered with TBS for me. 

Cheers, Hugo

andyp1per replied:
Hugo, you can use 0x81 but please provide some kind of frame description so that this can be put into the CRSF specification. Thx Andy

qqqlab replied:
The frame is a standard (not extended) CRSF frame, without any data formatting for the payload. Any payload data formatting is handled by the AP Lua and transmitter Lua scripts. It is basically a transparent link, like the 3DR/SiK Radios, and could theoretically be used for MavLink...

CRSF Frame Format
```
uint8_t device_addr
uint8_t frame_length
uint8_t type = 0x81
uint8_t payload[1..60]
uint8_t crc
```
I fixed AP_RCProtocol_CRSF::process_lua_pop() to exclude crc.

I also raised a PR with ExpressLRS to include the 0x81 frame, see: https://github.com/ExpressLRS/ExpressLRS/pull/2062 . However, this AP PR will work also without the ELRS PR implemented, by setting the first payload byte to 0x00 in crossfireTelemetryPush() on the transmitter side.

I hope this PR is good to go now ;-)

andyp1per commented on libraries/AP_RCProtocol/AP_RCProtocol_CRSF.cpp:
```diff
@@ -627,3 +628,14 @@ namespace AP {
 };
+
+//save data frame to buffer for lua, overwrite any existing buffer data
+void AP_RCProtocol_CRSF::process_lua_pop()
```
Needs to be surrouneded by #if AP_SCRIPTING_ENABLED

andyp1per commented on libraries/AP_RCProtocol/AP_RCProtocol_CRSF.h:
```diff
@@ -279,6 +279,14 @@ class AP_RCProtocol_CRSF : public AP_RCProtocol_Backend {
     // return the protocol string
     const char* get_protocol_string(ProtocolType protocol) const;
 
+    struct lua_buffer_t {
```
#if AP_SCRIPTING_ENABLED

andyp1per commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -434,6 +434,10 @@ void AP_CRSF_Telem::process_packet(uint8_t idx)
             calc_flight_mode();
             break;
         case PASSTHROUGH:
+            //first push lua data (if available), then other telem data
```
#if AP_SCRIPTING_ENABLED
need a custom frame id
there needs to be a way of switching this off. Please add an option bit to RC_OPTIONS, default off

qqqlab replied:
I don't think an additional RC_OPTIONS bit is required. The bidirectional telemetry option is switched off by default, only after the first Lua call to crsfPop/Push the pop/push buffers are allocated and the bidirectional telemetry becomes active.

yaapu replied:
Hi, I like the general idea, very nice!
but I'm afraid you won't get much bandwidth the way you designed it, you'll get a chance of sending a packet FC->TX only at "passthrough scheduler rate" which is quite low, around 10Hz, and not any faster (the scheduler has a rate limit) so your actual througput will be lower than expected, on ELRS we can probably go faster but the scheduler rates are designed to be compatible with both ELRS and CRSF.
Also we do not shut standard CRSF telemetry off so quite some bandwidth is used for legacy CRSF telemetry frames as well. 
Your code will starve passthrough packets and be delayed by other CRSF streams, a better approach IMO would be to declare an additional scheduler stream with a potentially much higher priority, this way you would get a bigger chance of sending packets without actually starving any stream since the packet scheduler is "fair" by design

qqqlab replied:
Hi yaapu, thanks! It was your excellent telemetry script that inspired me to start this PR, in order to eventually also get your ground station script to work with elrs.

I'm aware of the scheduler issues, but did not feel comfortable at this point to mess around with it. The code works as is, but there is definitely room for improvement!

qqqlab replied:
Wouldn't it be a clean solution to add Lua functions crsfSetFreq(Hz) and crsfStandardTele(on|off) to control this?

andyp1per commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -1400,6 +1404,22 @@ void AP_CRSF_Telem::calc_status_text()
     _statustext.available = false;
 }
 
+// Get packet of lua push telemetry data
```
#if AP_SCRIPTING_ENABLED

andyp1per reviewed (changes requested):
Very nice work! Needs a few changes.

IamPete1 commented on libraries/AP_Scripting/lua_bindings.cpp:
```diff
@@ -532,3 +535,65 @@ int lua_removefile(lua_State *L) {
+        lua_pushlstring(L, buf->data, buf->len);
+        buf->len = 0;
+    } else {
+        lua_pushnil(L);
+        lua_pushnil(L);
```
```suggestion
        return 0;
```

IamPete1 commented on libraries/AP_Scripting/lua_bindings.cpp:
```diff
@@ -532,3 +535,65 @@ int lua_removefile(lua_State *L) {
+
+// local cmd, data = crsf_pop() -- get byte array from crsf or nil if no data received
+int lua_crsf_pop(lua_State *L) {
+#if HAL_CRSF_TELEM_ENABLED
```
You can just def out the whole method.

IamPete1 replied:
We might need to mess with the generator to get it  to cope.

qqqlab replied:
Deffing out the whole method was also my first thought, but have no idea how the generator reacts.

IamPete1 replied:
doing a sub table should allow you to add a `depends` to the table. Again see `CAN`

IamPete1 commented on libraries/AP_Scripting/docs/docs.lua:
```diff
@@ -2382,3 +2382,16 @@ function dirlist(directoryname) end
+---@param cmd integer -- byte CRSF type
+---@param data string -- byte data to send
+---@return integer -- Number of bytes sent
+function crsfPush(cmd, data) end
```
I would prefer to add `crsf` as a table and then `push` as a function. So it would be `crsf:push(cmd, data)`

qqqlab replied:
Fine with me, but don't know how to do this.

IamPete1 replied:
See `CAN` bindings as a example.

IamPete1 commented on libraries/AP_Scripting/lua_bindings.cpp:
```diff
@@ -532,3 +535,65 @@ int lua_removefile(lua_State *L) {
+#if HAL_CRSF_TELEM_ENABLED
+    int push_len = 0;
+    int isnum;
+    uint8_t cmd = lua_tointegerx(L, 1, &isnum);
```
you can use the helper `get_uint8_t(L, 1)` It will not return if value is invalid.

IamPete1 commented on libraries/AP_Scripting/lua_bindings.cpp:
```diff
@@ -532,3 +535,65 @@ int lua_removefile(lua_State *L) {
+        AP_RCProtocol_CRSF::lua_buffer_t *buf = crsf->lua_push_buffer;
+        if (buf && buf->len == 0) {
+            WITH_SEMAPHORE(buf->sem);
+            push_len = (len < sizeof(AP_RCProtocol_CRSF::lua_buffer_t::data) ? len : sizeof(AP_RCProtocol_CRSF::lua_buffer_t::data));
```
Better to error out if the length is too long rather than just throwing away the end of the message.

IamPete1 replied:
Ah, I see, you return the number of bytes sent. But its a constant max lenght? Do we need that return?

qqqlab replied:
I agree, this can be done better. What do you think of this design?

data = crsf:pop()
data: nil=no data or alloc fail
(removed cmd return value because it is always 0x81 now)

result = crsf:push(data, [cmd=0x81])
result: true=ok, false=not pushed, nil=data too long or alloc fail
optional cmd argument

IamPete1 replied:
I think alloc fail is OK to error out on (because I don't know what the script could do to fix), maybe data too long as well? If you try and send too much isn't that just a programming error? Then you just get a bool back?

qqqlab replied:
Alloc fail is a hard fault, but is it worth to waste flash for unlikely error messages? Or is there a pre-fabricated "out of memory" message available?

The data too long behaviour I (sort of) mirrored from EdgeTX:

```
int luaCrossfireTelemetryPush(lua_State * L)
...
  if (lua_gettop(L) > TELEMETRY_OUTPUT_BUFFER_SIZE ) {
    lua_pushboolean(L, false);
    return 1;
  }
```

andyp1per commented:
Given this is essentially a passthrough blob for LUA I'm wondering whether this should be done as subtype of the regular ardupilot passthrough frame. Would lose a byte in the payload so there's a downside. I'm also very uncomfortable with there being no formatting at all - you are essentially relying on both ends being identical and written by the same person. What happens when you want to change something? You are going to break all consumers. So I think at a minimum you need some kind of typeid in the payload that you can use for version/payload type/whatever to prevent pain for future users.

yaapu commented on libraries/AP_Scripting/applets/ExpressLRS.md:
```diff
@@ -0,0 +1,87 @@
+
+Tested with: Arducopter 4.4.0-DEV + ExpressLRS 3.1.2 100Hz Full 1:2 + EdgeTX 2.8.0
+
+NOTE: the first data byte _NEEDS_ to be 0x00 or 0xc8 in EdgeTX `crossfireTelemetryPush(command, data)`, otherwise no data is received. (This appears to be a bug/feature in ExpressLRS 3.1.2 and/or EdgeTX 2.8.0)
```
this is not a bug, it's by design, 0x00 is a broadcast address and 0xC8 is flightcontroller device address, so in your testing you should target the flight controller, 0x00 is probably too broad

qqqlab commented:
Sorry guys, I disagree. My thought behind this PR is to create an efficient transparent link between Lua on the FC and Lua on the transmitter via elrs. Best analogue I can think of is: UDP, raw data is packetized but no guarantee that it will arrive.

Transparent means that all parties in between (FC, elrs, EdgeTx) do not have raw data formatting requirements, formatting is handled by the Lua scrips on both ends. Does this open a door to bad things to happen? Of course it does. But bad things can happen already now, as strict control over the FC was already lost by including Lua in the project: execute one wrong FC Lua statement and the vehicle crashes. So yes, better instruct Lua pilots to make use of a mechanism (for example a handshake) to enforce that both FC and TX Lua scripts are talking to correct recipients.

Efficient means not wasting bytes. The elrs over-the-air packet only has a 5 or 10 byte payload, and does not have retransmit capabilities. Sending long messages will quickly deteriorate the long range capabilities, because if one ota packet fails to arrive then the whole message is lost. With the proposed 0x81 type code the ota overhead is 3 bytes: dev_adr, len, type. 

Broadcast: I'm not sure we're talking about the same thing here. The crsf protocol defines two broadcast bytes: one for dev_adr (1st byte) and one for destination (4th byte). The dev_adr is rewritten by elrs on reception, so no broadcast is used here. In this context I don't see the need for the destination byte (4th crsf byte) as defined by the extended crsf header. I would expect that any receiver of crsf messages silently ignores messages with an unknown type (0x81). Also the current crsf type 0x80 implementation does not have a destination byte, it uses the sub_type as 4th byte. I've sent a PR to ExpressLRS to include the 0x81 code without destination byte.

yaapu commented:
> Broadcast: I'm not sure we're talking about the same thing here. The crsf protocol defines two broadcast bytes: one for dev_adr (1st byte) and one for destination (4th byte). The dev_adr is rewritten by elrs on reception, so no broadcast is used here. In this context I don't see the need for the destination byte (4th crsf byte) as defined by the extended crsf header. I would expect that any receiver of crsf messages silently ignores messages with an unknown type (0x81). Also the current crsf type 0x80 implementation does not have a destination byte, it uses the sub_type as 4th byte. I've sent a PR to ExpressLRS to include the 0x81 code without destination byte.

I was referring to the first byte of any header, I was suggesting to use 0x8c to specificaly target the flight controller from the TX->RX traffic, not the other way around

andyp1per commented:
@qqqlab this needs to work over regular CRSF as well (i.e. TBS kit). Regular CRSF will do routing based on the destination ids (for instance VTX traffic can be routed through a RX). We need to make sure this still works - i.e. an ELRS-only solution is no good. Did trappy reach out to you? I want to get this in the CRSF spec, but it will need his blessing to do so. I don't think UDP is a good analogy here, your implementation assumes specific source and target (LUA), which means some kind of negotiation. Still think you need some kind of id byte to make sure no-one gets burned.

qqqlab commented:
@andyp1per 
Reading the docs: //Extended Header Frames, range: 0x28 to 0x96

So, the CRSF header is: `type dest origin`

But in AP the 0x80 telemetry frame is: `type sub_type payload`

This confused me into thinking that `dest` can be part of the payload. The trick is that the possible`sub_type` values 0xF0-F2 are not valid CRSF destinations. And because of that, `origin` can be part of the payload.

New proposal: use `0x80 0xF3 payload` as CRSF package, does that sound reasonable to you?

qqqlab commented:
@andyp1per Thinking further, I propose the following solution. This should work with any CRSF equipment, as it sets the correct CRSF destination values:

FC to RC: `0x80 0xEA payload`
RC to FC: `0x80 0xEC payload`

in addition to the existing:

FC to RC: `0x80 0xF0-F2 payload`

CRSF_FRAMETYPE_AP_CUSTOM_TELEM = 0x80,
CRSF_ADDRESS_RADIO_TRANSMITTER = 0xEA,
CRSF_ADDRESS_CRSF_RECEIVER = 0xEC,
CRSF_AP_CUSTOM_TELEM_SINGLE_PACKET_PASSTHROUGH = 0xF0,
CRSF_AP_CUSTOM_TELEM_STATUS_TEXT = 0xF1,
CRSF_AP_CUSTOM_TELEM_MULTI_PACKET_PASSTHROUGH = 0xF2,

This solution leaves space for future expansion of the CRSF_FRAMETYPE_AP_CUSTOM_TELEM type by adding additional CRSF_AP_CUSTOM_TELEM_xxx constants.

andyp1per commented:
@qqqlab so you have reinvented extended header frames 😄 I have wondered why the passthrough stuff is not an extended header frame, but that ship has sailed...and now I have re-read the spec, TBS is expecting the ArduPilot frames to be extended header frames but we have not implemented them that way - @yaapu do you know why we did it that way?  An extended header frame looks like this:

![image](https://user-images.githubusercontent.com/2893260/216116221-029ef6c2-1204-420a-84b7-4e3e6442185e.png)

I don't think your proposal is unreasonable, but I wonder whether it is necessary. We could make 0x81 a regular extended header frame, then all of the addressing info is included. We could also just stick with a 0x80 frame with a subtype, this subtype could then be changed for newer versions, but I suspect what you really want is to put the onus on the LUA code to write a version byte as the first byte of the payload so that scripts can accept or reject based on compatibility. This is LUA, so I imagine someone will write an alternative to your script quite soon - how do you stop your code blowing up and vice version, some kind of version check seems necessary.

Broadcast frames are supposed to be directionless.

There is a CRSF spec which Alex and I have access to, but its not public. To be clear I won't accept this PR into the CRSF code unless it is compatible with CRSF as a whole (i.e. CRSF + ELRS), but I don't think that is a high bar, it should be easy to accomplish - just need to make sure that we get this right. Let me have a chat to Trappy and see if he has some good suggestions here.

yaapu commented:
> @yaapu do you know why we did it that way

No, I'd assume I/we overlooked this detail, but I think we can fix it, opentx/edgetx/ethos should not make assumptions on frame header types so they should politely pass even extended header frames up

qqqlab commented:
The CRSF frame is sent 1-to-1 over the air, so adding more headers bytes is not a good idea, it will kill the throughput very quickly. One full ELRS frame has only 10 bytes of payload, of which 6 are eaten by the extended header. So even short messages will be transmitted in multiple OTA frames, increasing the risk of lost messages. Other long range RC radio systems (TBS?) will not do much better, because longer frames will reduce frame rate or reduce reception distance. It is one or the other, you can't have it all...

In short: that's why I was hoping that there would be a CRSF type with standard headers available, reducing the overhead from 6 to 4 bytes (sync,len,type,crc). But, the current solution is a compromise that works fine for me: (ab)using the origin byte as the first payload byte should not hurt the CRSF ecosystem. No need for a new extended CRSF type.

Yes, I very much want to put the onus on Lua. The link is a pipe. There is no need to send version bytes in every frame if the Lua scripts use a handshake. For example: TX sends "hello, do you speak DRONELUA44?", RX answers "yes, I speak DRONELUA44", and after that the two can communicate in DRONELUA44. If the RX does not understand the initial hello message, it won't be able to send the correct answer back, and both sides know: this is not going to work.

Anyway, I just pushed an update, these are the changes:
 - Scripts commands are now crsf:pop() and crsf:push()
 - The script handlers now have fancy error messages, at the cost of 64 additional flash bytes.
 - Removed 0x81 CRSF type
 - Using 0x80 0xC8 (FC) and 0x80 0xEA (Radio Transmitter) as message headers instead (mentioned wrong code in my previous comment)
 - Placed the transmitting code in AP_CRSF_Telem::_get_telem_data()

@yaapu I'm trying to give lua messages the highest priority, and use the existing scheduler when no lua message is available. My current attempt is in AP_CRSF_Telem::_get_telem_data(), but the result is that both telemetry and lua rates are way too low. If I bypass the scheduler completely by adding "return false;" before "run_wfq_scheduler();" it works fine, but of course without telemetry. Any ideas on this?

qqqlab commented:
@yapuu Just read your previous post, I'll give it a try. If you have other suggestions, please let me know.

> Your code will starve passthrough packets and be delayed by other CRSF streams, a better approach IMO would be to declare an additional scheduler stream with a potentially much higher priority, this way you would get a bigger chance of sending packets without actually starving any stream since the packet scheduler is "fair" by design

yaapu commented:
> > @yaapu do you know why we did it that way
> 
> No, I'd assume I/we overlooked this detail, but I think we can fix it, opentx/edgetx/ethos should not make assumptions on frame header types so they should politely pass even extended header frames up

The main reason was that I added it to the "telemetry" frames type, since that's what basically is, but it somehow is out of specs unfortunately.
I talked with the elrs team, if we switch to extended frames we would brake their inspection of the first payload byte looking for the status text subtype id needed to queue those frames an a dedicated buffer, alternative would be perhaps to define an additional frame type dedicated to status text messages because if not handled with a dedicated queue any telemetry frame would overwrite them, elrs has a 1 item buffer per frame type

yaapu commented:
> @yaapu I'm trying to give lua messages the highest priority, and use the existing scheduler when no lua message is available. My current attempt is in AP_CRSF_Telem::_get_telem_data(), but the result is that both telemetry and lua rates are way too low. If I bypass the scheduler completely by adding "return false;" before "run_wfq_scheduler();" it works fine, but of course without telemetry. Any ideas on this?

it is tricky, indeed. I guess that bypassing run_wfq_scheduler() is a no go, you would cut too much, what we do now for similar use cases is dynamically change stream priority or shutting individual streams on/off based on context.

yaapu commented:
> * Removed 0x81 CRSF type
> * Using 0x80 0xC8 (FC) and 0x80 0xEA (Radio Transmitter) as message headers instead (mentioned wrong code in my previous comment)

this would allow CRSF/ELRS to treat lua frames as extended frames, should work just fine, the origin byte will most likely be unmappable, that's the way it is now so btw

qqqlab commented:
I changed the title of this PR to _"Lua CRSF Interface"_, to better reflect it's purpose. This PR exposes the CRSF port to Lua for reading and writing, without any requirements about what type of CRSF equipment is connected. ExpressLRS bidirectional telemetry is just one of the many possible use-cases.

Changes in the new commit:
 - Added Lua crsf:telemetry_enabled(true|false) to enable/disable Ardupilot CRSF Telemetry
 - Concurrent Lua Telemetry & AP Telemetry works
 - Updated the Lua examples to use extended CRSF frames

I did a day of testing with ExpressLRS and it works fine. I'm not able to test with other CRSF equipment, but expect that it will also work.

Looking forward to your comments.

andyp1per commented:
Wasn't able to talk to Trappy today - hopefully tomorrow. Thanks for all your work on this @qqqlab

andyp1per commented:
Alex and I have lots of TBS kit we can test on.

Just one comment - handshaking only works if both sides understand the handshake, if one end is completely different then things will just break. I understand the desire to save as much payload space as possible, but needs to not be at the cost of good behaviour for the majority of users.

yaapu commented:
I have an idea for the extended frame issue, how about asking TBS to register 0xF0-0xF2 as custom ardupilot destinations, we could keep the protocol almost the way it is and elrs integration would still work just fine and TBS just as well, routing could be done based on the destinatin which would be documented by TBS specs. IMO payload should NOT be shifted by 1 byte and we should simply reuse the origin byte as payload to save some space

qqqlab committed:
Merge branch 'ArduPilot:master' into pr/csrf_pop

qqqlab commented:
@andyp1per Hi Andy, I've made the changes you requested and did additional testing. I hope you can review the code again.

 - add new demo Lua programs, see readme.md
 - add filter to only forward to Lua extended header CRSF messages with destination 0xC8 or 0x00
 - add Lua function to switch off AP CRSF Telemetry. I tried to do this by setting RC_OPTIONS bit 8 = 0, but even with this bit off AP still is quite chatty on CRSF: it sends CRSF telemetry messages like VARIO 0x07, GPS 0x02, ATTITUDE 0x1E

qqqlab committed:
Merge branch 'master' into pr/csrf_pop

qqqlab committed:
Update docs.lua

qqqlab committed:
Update CRSF-Telem.lua

qqqlab committed:
Update CRSF-Debug.lua

qqqlab committed:
Update CRSF-GCS-E.lua

qqqlab committed:
Update CRSF-Rate-E.lua

qqqlab committed:
Merge branch 'ArduPilot:master' into pr/csrf_pop

qqqlab committed:
Update docs.lua

peterbarker commented:
Ping on this one - have things settled down in whatever external arguments were being had?

Is this still a good thing to proceed with?
